nixos/lldap: bootstrap `jwt_secret` if not provided, lldap: remove emilylange from maintainers

emilylange opened the pull request:
## Description of changes

If `jwt_secret` is not provided, `lldap` defaults to `secretjwtsecret` as value which is hardcoded in the code base.

See https://github.com/lldap/lldap/blob/v0.5.0/server/src/infra/configuration.rs#L76-L77

This is really bad, because it is trivially easy to generate an admin access token/cookie as attacker, if a `jwt_secret` is known.

Also, I will drop maintainership of lldap.
I find lldap's defaults security-wise and its security-posture in a broader sense deeply unsettling for something as security-critical an authentication server.

This is not a breaking change.
But it will invalidate existing sessions, which I consider fine.

## Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [x] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.05 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) (or backporting [23.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md) and [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

emilylange committed:
nixos/lldap: bootstrap `jwt_secret` if not provided

If not provided, lldap defaults to `secretjwtsecret` as value which is
hardcoded in the code base.

See https://github.com/lldap/lldap/blob/v0.5.0/server/src/infra/configuration.rs#L76-L77

This is really bad, because it is trivially easy to generate an admin
access token/cookie as attacker, if a `jwt_secret` is known.

emilylange committed:
lldap: remove emilylange from maintainers

I find lldap's defaults security-wise and its security-posture in a
broader sense deeply unsettling for something as security-critical an
authentication server.

bjornfor commented:
> This is really bad, because it is trivially easy to generate an admin access token/cookie as attacker, if a jwt_secret is known.

Is there an upstream issue about it?

bjornfor reviewed (approved):
LGTM.

bjornfor commented on nixos/modules/services/databases/lldap.nix:
```diff
@@ -107,8 +107,17 @@ in
       wants = [ "network-online.target" ];
       after = [ "network-online.target" ];
       wantedBy = [ "multi-user.target" ];
+      script = lib.optionalString (!cfg.settings ? jwt_secret) ''
```
You're checking two env vars below, but only one attr in `settings` above -- is that correct? (There is no *file option in `cfg.settings`?)

emilylange replied:
Yes. From what I know `_FILE` is only available for environment variables and are invalid in the `configuration.toml`.

MinerSebas commented on nixos/modules/services/databases/lldap.nix:
```diff
@@ -107,8 +107,17 @@ in
       after = [ "network-online.target" ];
       wantedBy = [ "multi-user.target" ];
+      script = lib.optionalString (!cfg.settings ? jwt_secret) ''
+        if [[ -z "$LLDAP_JWT_SECRET_FILE" ]] && [[ -z "$LLDAP_JWT_SECRET" ]]; then
```
This should have a comment explaining the Problem.

emilylange replied:
Yes, good idea.
Pushed a comment a few moments ago :)

emilylange commented:
> > This is really bad, because it is trivially easy to generate an admin access token/cookie as attacker, if a jwt_secret is known.
> 
> Is there an upstream issue about it?

Not that I know of.

https://github.com/lldap/lldap/blob/v0.5.0/README.md#with-docker mentions multiple times that `jwt_secret` should be set and the latest version on the `main` branch also mentions and provides https://github.com/lldap/lldap/blob/533d1bcfd0189581c299b59c8120860d1e6f0e89/generate_secrets.sh as helper.

And there is also a start-up warning when using the default (hardcoded) value.

We, on our side, mention `LLDAP_JWT_SECRET_FILE` and `LLDAP_LDAP_USER_PASS_FILE` in the `services.lldap.environment` example, but I, being the author of that module, recently realized that most end users probably won't notice and don't expect a hardcoded default.

We cannot catch this at eval time, because one might use `services.lldap.environmentFile` to set `LLDAP_JWT_SECRET`.
Bootstrapping the secret in runtime if it's missing looks to me like the way to go here.

But I am also open for different solutions.
I just felt like providing one final fix for this issue before removing myself from lldap's maintainers.

bjornfor commented on nixos/modules/services/databases/lldap.nix:
```diff
@@ -107,8 +107,21 @@ in
+      script = lib.optionalString (!cfg.settings ? jwt_secret) ''
+        if [[ -z "$LLDAP_JWT_SECRET_FILE" ]] && [[ -z "$LLDAP_JWT_SECRET" ]]; then
+          if [[ ! -e "./jwt_secret_file" ]]; then
+            ${lib.getExe pkgs.openssl} rand -base64 -out ./jwt_secret_file 32
```
Is this running with custom umask at the service level? If not I think it will create jwt_secret_file world readable... Perhaps do `(umask 027; openssl ...)`?

emilylange replied:
systemd's `DynamicUser=true` does some heavily lifting for us here.

`/var/lib/lldap` is just a symlink to `/var/lib/private/lldap`, while `/var/lib/private` has `0700/drwx------` perms, making it inaccessible to any unprivileged users.

But given there is also #287873, this is a really good and important catch nonetheless -- will fix. Thank you!

Also went ahead and checked what the perms of other files in `/var/lib/lldap` are (`nixosTests.lldap`).
```
$ ls -lash /var/lib/lldap/
total 128K
4.0K drwxr-xr-x 2 lldap lldap 4.0K Mar 10 16:57 .
4.0K drwx------ 3 root  root  4.0K Mar 10 16:57 ..
4.0K -rw-r--r-- 1 lldap lldap   45 Mar 10 16:57 jwt_secret_file
4.0K -r-------- 1 lldap lldap  128 Mar 10 16:57 server_key
112K -rw-r--r-- 1 lldap lldap 108K Mar 10 16:57 users.db
```

That's not great.
I would argue the `users.db` shouldn't be `0644/-rw-r--r--`.

Given this, I prefer service level `UMask=027` so the `umask` not only applies to `jwt_secret_file` but also to `users.db` and potentially other files as well.

And for instances that already have `users.db` set as is, I think changing `StateDirectoryMode=` from its default `0755` to `0750` should do the job.

emilylange committed:
nixos/lldap: set service `UMask=0027` and `StateDirectoryMode=0750`

While `/var/lib/lldap` isn't technically accessible by unprivileged
users thanks to `DynamicUser=true`, a user might prefer and change it to
`DynamicUser=false`.

There is currently also a PR open that intends to make `DynamicUser`
configurable via module option.

As such, `jwt_secret_file`, if bootstrapped by the service start
procedure, might be rendered world-readable due to its permissions
(`0644/-rw-r--r--`) defaulting to the service's umask (`022`) and
`/var/lib/lldap` to `0755/drwxr-xr-x` due to `StateDirectoryMode=0755`.

This would usually be fixed by using `(umask 027; openssl ...)` instead
of just `openssl ...`.

However, it was found that another file (`users.db`), this time
bootstrapped by `lldap` itself, also had insufficient permissions
(`0644/-rw-r--r--`) inherited by the global umask and would be left
world-readable as well.

Due to this, we instead change the service's to `027`.

And to lower the impact for already bootstrapped files on existing
instances like `users.db`, set `StateDirectoryMode=0750`.

bjornfor commented:
@emilylange: Thanks!
